movement/io/validators.py

```diff
@@ -1,39 +1,348 @@
-import logging
+import os
 from pathlib import Path
+from typing import Any, Iterable, List, Literal, Optional, Union
 
-from pydantic import BaseModel, field_validator
+import h5py
+import numpy as np
+from attrs import converters, define, field, validators
 
-# initialize logger
-logger = logging.getLogger(__name__)
+from movement.logging import log_and_raise_error, log_warning
 
 
-class DeepLabCutPosesFile(BaseModel):
-    """Pydantic class for validating files containing
-    pose estimation results from DeepLabCut (DLC).
+@define
+class ValidFile:
+    """Class for validating file paths.
 
-    Pydantic will enforce the input data type.
-    This class additionally checks that the file exists
-    and has a valid suffix.
+    Parameters
+    ----------
+    path : str or pathlib.Path
+        Path to the file.
+    expected_permission : {'r', 'w', 'rw'}
+        Expected access permission(s) for the file. If 'r', the file is
+        expected to be readable. If 'w', the file is expected to be writable.
+        If 'rw', the file is expected to be both readable and writable.
+        Default: 'r'.
+    expected_suffix : list of str
+        Expected suffix(es) for the file. If an empty list (default), this
+        check is skipped.
+
+    Raises
+    ------
+    IsADirectoryError
+        If the path points to a directory.
+    PermissionError
+        If the file does not have the expected access permission(s).
+    FileNotFoundError
+        If the file does not exist when `expected_permission` is 'r' or 'rw'.
+    FileExistsError
+        If the file exists when `expected_permission` is 'w'.
+    ValueError
+        If the file does not have one of the expected suffix(es).
+    """
+
+    path: Path = field(converter=Path, validator=validators.instance_of(Path))
+    expected_permission: Literal["r", "w", "rw"] = field(
+        default="r", validator=validators.in_(["r", "w", "rw"]), kw_only=True
+    )
+    expected_suffix: List[str] = field(factory=list, kw_only=True)
+
+    @path.validator
+    def path_is_not_dir(self, attribute, value):
+        """Ensures that the path does not point to a directory."""
+        if value.is_dir():
+            log_and_raise_error(
+                IsADirectoryError,
+                f"Expected a file path but got a directory: {value}.",
+            )
+
+    @path.validator
+    def file_exists_when_expected(self, attribute, value):
+        """Ensures that the file exists (or not) depending on the expected
+        usage (read and/or write)."""
+        if "r" in self.expected_permission:
+            if not value.exists():
+                raise log_and_raise_error(
```

Comment:
Oh my bad, that "raise" should not be there! I'm surprised the linters didn't catch this.

Reply:
I will go with your first suggestion here.
I'll rename `log_an_raise_error()` to `logged_error()` and have it return the Expetion instead of raising it. This way we can do things like `raise logged_error(ValueError, "error message") from e`

Reply:
Sounds good. I think we want ``log_error()`` to be consistent with ``log_warning()``